### PR TITLE
Use GitHub Actions to publish registry

### PR DESCRIPTION
**Pull request by jablko**
Can we use GitHub Actions to publish the [types-registry](https://www.npmjs.com/package/types-registry) on a weekly schedule, vs. checking each time we run the publisher, if it's been seven days yet? https://github.com/microsoft/DefinitelyTyped-tools/blob/3cbaae228b0fb417f8cf86a7e094174922949fa4/packages/publisher/src/publish-registry.ts#L80

What I did was:
- Add a "Publish registry" workflow that runs the [parse](https://github.com/microsoft/DefinitelyTyped-tools/tree/master/packages/publisher#parse-the-definitions), [calculate-versions](https://github.com/microsoft/DefinitelyTyped-tools/tree/master/packages/publisher#calculate-versions), and [publish-registry](https://github.com/microsoft/DefinitelyTyped-tools/tree/master/packages/publisher#calculate-versions) scripts.
- In each script, replace `defaultLocalOptions` with `defaultRemoteOptions` if [`GITHUB_ACTIONS`](https://docs.github.com/en/actions/learn-github-actions/environment-variables#default-environment-variables) is true: https://github.com/microsoft/DefinitelyTyped-tools/blob/3cbaae228b0fb417f8cf86a7e094174922949fa4/packages/publisher/src/publish-registry.ts#L50
- In the "full" script, skip publish-registry so we no longer maybe-publish it on each run.
- Remove the seven day condition from publish-registry, deferring instead to the workflow schedule.

Here's an example of it working (except for actually publishing to npm): https://github.com/jablko/DefinitelyTyped-tools/actions/runs/2271150100

### CODE REVIEW

**Review by sandersn — CHANGES_REQUESTED** [review]
Looks good -- a lot of questions about the yml since I don't know the actions ecosystem very well.

File: .github/workflows/publish-registry.yml [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+      - uses: actions/checkout@v3
+      - uses: actions/setup-node@v3
+        with:
+          cache: yarn
+      - run: yarn install --frozen-lockfile
+      - run: yarn build
+      - name: Parse declarations
+        run: yarn workspace @definitelytyped/publisher parse
+      # https://github.com/microsoft/types-publisher/commit/7fa7f9c3cbd0c8bd6131398b4fea1faf9baf9665#diff-c4063aea1f6d1c21c51407252387627e9a506c735f8dbd58fbf82ae821fae128R170
+      - uses: actions/cache@v3
```
[the remark this answers is not in the document]
Comment (jablko) [reply .github/workflows/publish-registry.yml:20]: Yes to everything except, if we cache `npmInfo.json` then we only hit the npm registry for `@types` packages that changed in the past week, which is faster than fetching every `@types` packages' metadata. Calculate versions only [revalidates `@types` packages if they changed](https://github.com/microsoft/types-publisher/blob/e57e0ca78d167165a9d19b954a94b6e12488e1d2/src/lib/versions.ts#L124-L128).

File: .github/workflows/publish-registry.yml [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+          key: cache-${{ github.run_id }}
+          restore-keys: cache-
+      - name: Calculate versions
+        run: yarn workspace @definitelytyped/publisher calculate-versions
+      - name: Publish registry
+        run: yarn workspace @definitelytyped/publisher publish-registry
+        env:
+          NPM_TOKEN: ${{ secrets.NPM_RETAG_TOKEN }}
+      - if: always()
+        uses: actions/upload-artifact@v3
```
[the remark this answers is not in the document]
Comment (jablko) [reply .github/workflows/publish-registry.yml:32]: Like the logs they help with observing what the publisher is doing internally. You can inspect them like you inspect the logs, [from the workflow run UI](https://github.com/jablko/DefinitelyTyped-tools/actions/runs/2271150100).

Without them you can try to reproduce the state of the run locally and recreate the files that way, but getting at the actual files saves time.

They're stored ["on GitHub"](https://docs.github.com/en/actions/using-workflows/storing-workflow-data-as-artifacts#about-workflow-artifacts) and are subject to the same retention policy as the logs.

File: .github/workflows/publish-registry.yml [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+          path: packages/publisher/cache/
+          key: cache-${{ github.run_id }}
+          restore-keys: cache-
+      - name: Calculate versions
+        run: yarn workspace @definitelytyped/publisher calculate-versions
+      - name: Publish registry
+        run: yarn workspace @definitelytyped/publisher publish-registry
+        env:
+          NPM_TOKEN: ${{ secrets.NPM_RETAG_TOKEN }}
+      - if: always()
```
[the remark this answers is not in the document]
Comment (jablko) [reply .github/workflows/publish-registry.yml:31]: [Yup, exactly](https://docs.github.com/en/actions/learn-github-actions/expressions#always), so you can inspect those files (the intermediate states) on success or failure, like the logs.

File: .github/workflows/publish-registry.yml [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+        with:
+          path: packages/publisher/cache/
+          key: cache-${{ github.run_id }}
+          restore-keys: cache-
+      - name: Calculate versions
+        run: yarn workspace @definitelytyped/publisher calculate-versions
+      - name: Publish registry
+        run: yarn workspace @definitelytyped/publisher publish-registry
+        env:
+          NPM_TOKEN: ${{ secrets.NPM_RETAG_TOKEN }}
```
[the remark this answers is not in the document]
Comment (jablko) [reply .github/workflows/publish-registry.yml:30]: It belongs to the [`types`](https://www.npmjs.com/~types) npm user, which is the user that publishes the `types-registry`:
```sh
npm view types-registry _npmUser
types <ts-npm-types@microsoft.com>
```

`secrets.NPM_TOKEN` on the other hand belongs to the [`definitelytyped-publisher`](https://www.npmjs.com/~definitelytyped-publisher), which is the one that publishes the DT tools --- `@definitelytyped/dtslint`, etc.:
```sh
npm view @definitelytyped/dtslint _npmUser
definitelytyped-publisher <dt-publisher@microsoft.com>
```

So for the `types-registry`, `secrets.NPM_RETAG_TOKEN` is what we want.

**Pull request merged**